Refactor gpg module and add ability to trust a key from URL

kwryankrattiger opened the pull request:
Often build caches will host their public keys in a common location to be downloaded. This allows users to use spack to download and trust keys directly rather than as two steps.

In addition to URL support, also refactor the gpg module in preparation for better localized gpg agent control. This is going to be nice to have to better support multiple keyrings for signing/managing keys.

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

kwryankrattiger committed:
Allow downloading/trusting a key from url

Often build caches will host their public keys in a common location to
be downloaded. This allows users to use spack to download and trust keys
directly rather than as two steps.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Allow downloading/trusting a key from url

Often build caches will host their public keys in a common location to
be downloaded. This allows users to use spack to download and trust keys
directly rather than as two steps.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

haampie commented:
Hm, there should be a manual confirmation step for trusting a key. I don't think we should add this before fixing the issues with "trusting" keys in Spack.

kwryankrattiger committed:
Add confirmation prompt before importing key

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Add confirmation prompt before importing key

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
CI: auto-trust keys when running headless CI commands

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
CI: auto-trust keys when running headless CI commands

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Update completions

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Update completions

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Add more auto-trust for tests

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Add more auto-trust for tests

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

haampie commented on lib/spack/spack/util/gpg.py:
```diff
@@ -230,6 +238,18 @@ def export_keys(location, keys, secret=False):
+    """Extract the public key ids from a file
+
+    Args:
+        keyfile (str): file with the public key
```
move (str) to function signature

kwryankrattiger replied:
copy-paste, will fix the other one too

haampie commented on lib/spack/spack/util/gpg.py:
```diff
@@ -230,6 +238,18 @@ def export_keys(location, keys, secret=False):
+    """
+    # Get the public keys we are about to import
+    output = GPG("--with-colons", keyfile, output=str, error=str)
+    return zip(_get_unimported_public_keys(output), _get_key_uids(output))
```
This makes me a bit uncomfortable. I'd rather see the `output` parsed once so we know the lhs and rhs of what's now being zipped corresponds to each other.

kwryankrattiger replied:
yeah I was thinking the same thing after writing it, I don't think it is possible that this is not 1:1 but it is also hard to read.

kwryankrattiger replied:
ok, it parses _almost_ everything now. I restrained myself from re-writing anymore of this module

kwryankrattiger committed:
GPG parser refactor to avoid multiple passes

Fully parse GPG output into objects so all information is available
after a single pass. Handles all of the current use cases and spack and
a few new ones that are still unused.

Fixed tests that mocked invalid gpg output.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
GPG parser refactor to avoid multiple passes

Fully parse GPG output into objects so all information is available
after a single pass. Handles all of the current use cases and spack and
a few new ones that are still unused.

Fixed tests that mocked invalid gpg output.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Spack requires gpg2 anyway, so this should always be true

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Spack requires gpg2 anyway, so this should always be true

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
WIP: More refactor

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Update completions

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger commented:
Closing in favor of https://github.com/spack/spack/pull/52430
